### PR TITLE
More unit tests for saving cart

### PR DESCRIPTION
**Pull request by dydome**
related to #11331

### CODE REVIEW

File: feature-libs/cart/saved-cart/components/add-to-saved-cart/add-to-saved-cart.component.spec.ts [hunk 1]
```diff
@@ -112,6 +127,20 @@ describe('AddToSavedCartComponent', () => {
     });
   });
 
+  describe('when user is not logged in', () => {
```
Comment (bgambocjaviniar) [inline feature-libs/cart/saved-cart/components/add-to-saved-cart/add-to-saved-cart.component.spec.ts]: This is already tested above. 

You tested the functionality for saving the cart, and that it can open a dialog from the component.ts.
What isn't available in the component, but is available in the template, which is the 'Saved Carts' button should be tested as well. Why we're testing it? Because it's one of the key parts of this component as it will redirect you to the listing page or login page.

- use case: when user is not logged in, he should be redirected to login page
- use case: when user is logged in, he should be redirected to the listing page

Screen Shot 2021-03-09 at 10 46 42 PM

Basically, what you do is, query the dom to get the element. Perform an event click, and expect what should be called.

Comment (bgambocjaviniar) [reply feature-libs/cart/saved-cart/components/add-to-saved-cart/add-to-saved-cart.component.spec.ts]: Remove the TODO comment afterwards

File: feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.html [hunk 2]
```diff
@@ -177,6 +181,7 @@ <h3 class="cx-saved-cart-form-title">
             </button>
             <button
               (click)="saveOrEditCart(cart?.code)"
+              [disabled]="form.invalid || isLoading$ | async"
```
Comment (bgambocjaviniar) [inline feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.html]: wrap the isLoading and pipe within an () 

Angular is complaining as it does not know what to do with the case you have above
```
form.invalid || (isLoading$ | async)
```

You can view the error by clicking the save cart for later or the edit icon. It will throw it in the devtools console

File: feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.ts [hunk 3]
```diff
@@ -151,16 +155,25 @@ export class SavedCartFormDialogComponent implements OnInit, OnDestroy {
...
+    form.setControl(
+      'description',
+      new FormControl('', [Validators.maxLength(this.descriptionMaxLength)])
+    );
+    this.form = form;
+    this.patchData(cart);
+  }
+
+  protected patchData(item?: any): void {
+    this.form.patchValue({ ...item });
```
Comment (bgambocjaviniar) [inline feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.ts:176]: if the name and description exist already, then the values will be populated in the cart page's dialog as well. You can test this by restoring a cart with a name and description set

Should we not display them when we're saving the cart?
I don't know the answer to be honest, but I think it shouldn't be there.

File: feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.spec.ts [hunk 4]
```diff
@@ -80,5 +113,128 @@ describe('SavedCartFormDialogComponent', () => {
...
+      expect(savedCartService.saveCart).toHaveBeenCalledWith({
+        cartId: mockCartId,
+        saveCartName: '',
+        saveCartDescription: '',
+        extraData: { edit: false },
+      });
+    });
+  });
+
+  describe('on delete cart', () => {
```
Comment (bgambocjaviniar) [inline feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.spec.ts]: You can just put them all in one `it`, which calls the deleteCart and expects those things to get triggered

File: feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.spec.ts [hunk 5]
```diff
@@ -80,5 +113,128 @@ describe('SavedCartFormDialogComponent', () => {
...
+    spyOn(launchDialogService, 'closeDialog');
+    component.close(mockSuccessDeleteCloseReason);
+
+    expect(launchDialogService.closeDialog).toHaveBeenCalledWith(
+      mockSuccessDeleteCloseReason
+    );
+  });
+
+  it('should return actual characters left', () => {
+    component?.form?.get('description')?.setValue('test');
```
Comment (bgambocjaviniar) [inline feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.spec.ts]: test with empty string and undefined as well

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]

File: feature-libs/cart/saved-cart/components/add-to-saved-cart/add-to-saved-cart.component.spec.ts (hunk 1 above)
Comment (dydome) [reply feature-libs/cart/saved-cart/components/add-to-saved-cart/add-to-saved-cart.component.spec.ts]: As I spoke with other guys, we should test it in e2e test not in units. So leaving this PR without this part.

Comment (bgambocjaviniar) [reply feature-libs/cart/saved-cart/components/add-to-saved-cart/add-to-saved-cart.component.spec.ts]: Agree

File: feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.ts (hunk 3 above)
Comment (bgambocjaviniar) [reply feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.ts:176]: Bill said we can leave the prefill form values on the save cart form (from cart page) as long it was a restored cart, therefore all is fine

File: feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.spec.ts [hunk 6]
```diff
@@ -80,5 +113,133 @@ describe('SavedCartFormDialogComponent', () => {
...
+      component?.form?.get('description')?.setValue('');
+
+      expect(component.descriptionsCharacterLeft).toEqual(
+        mockDescriptionMaxLength
+      );
+    });
+  });
+
+  //  TODO
+  // describe('watch save cart process success', () => {
```
Comment (bgambocjaviniar) [inline feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.spec.ts]: I left the comment in the DM regarding this to not forget this

Didn't we agree, that we will test this from the call on Tuesday?

Comment (bgambocjaviniar) [reply feature-libs/cart/saved-cart/components/saved-cart-form-dialog/saved-cart-form-dialog.component.spec.ts]: Someone can approve and you can merge it in afterwards

**Review by bgambocjaviniar — APPROVED** [review]
